pkg/server/endpoints/eventTracker.go

```diff
@@ -0,0 +1,171 @@
+	if pollPeriods < 1 {
+		pollPeriods = 1
+	}
+
+	// cleanup boundaries into incrasing slice of no duplicates
+	boundaryMap := make(map[uint]bool)
+	filteredBounds := []uint{}
+	for _, boundary := range boundaries {
+		// trim duplicates and boundaries outside of polling range
+		if _, found := boundaryMap[boundary]; !found && boundary < pollPeriods {
+			boundaryMap[boundary] = true
+			filteredBounds = append(filteredBounds, boundary)
+		}
+	}
+	slices.Sort(filteredBounds)
+
+	initialPolls := uint(0)
+	switch {
+	case boundaries == nil, len(filteredBounds) == 0:
+		initialPolls = pollPeriods
+	default:
+		initialPolls = filteredBounds[0]
+	}
+
+	return &eventTracker{
+		initialPolls: initialPolls,
+		pollPeriods:  pollPeriods,
+		boundaries:   filteredBounds,
+		events:       make(map[uint]*eventStats),
+	}
+}
+
+func (et *eventTracker) PollPeriods() uint {
+	return et.pollPeriods
+}
+
+func (et *eventTracker) InitialPolls() uint {
+	return et.initialPolls
+}
+
+func (et *eventTracker) PollBoundaries() []uint {
+	return et.boundaries
+}
+
+func (et *eventTracker) Polls() uint {
+	return et.initialPolls + uint(len(et.boundaries))
+}
+
+func (et *eventTracker) StartTracking(event uint) {
+	et.events[event] = &eventStats{
+		hash: hash(event),
+		ticks: 0,
+		polls: 0,
+	}
+}
+
+func (et *eventTracker) StopTracking(event uint) {
+	delete(et.events, event)
+}
+
+
+func (et *eventTracker) SelectEvents() []uint {
+	// fmt.Print("polling events\n")
+	pollList := make([]uint, 0)
+	for event, _ := range et.events {
+		eventStats := et.events[event]
+		bucket := eventStats.polls - et.initialPolls
+		// fmt.Printf("  event %d: %+v, bucket %d\n", event, eventStats, bucket)
+		switch {
+		case eventStats.polls < et.initialPolls:
+			// fmt.Print("  initial poll range, adding\n")
+			pollList = append(pollList, event)
+			eventStats.polls++
+		case bucket + 1 < uint(len(et.boundaries)):
+			// fmt.Print("  not last range\n")
+			bucketWidth := et.boundaries[1+bucket] - et.boundaries[bucket]
+			bucketPosition := eventStats.hash % bucketWidth
+			//fmt.Printf("event %d, hash %d, bucket %d\n", event, eventStats.hash, bucketPosition)
+			if eventStats.ticks == et.boundaries[bucket] + bucketPosition {
+				pollList = append(pollList, event)
+			}
+		case bucket < uint(len(et.boundaries)):
+			// fmt.Print("  last range\n")
+			bucketWidth := et.pollPeriods - et.boundaries[bucket]
+			bucketPosition := eventStats.hash % bucketWidth
+			//fmt.Printf("event %d, hash %d, bucket %d\n", event, eventStats.hash, bucketPosition)
+			if eventStats.ticks == et.boundaries[bucket] + bucketPosition {
+				pollList = append(pollList, event)
+			}
+		}
+		eventStats.ticks++
+	}
+	return pollList
+}
+
+func (et *eventTracker) EventCount() uint {
+	return uint(len(et.events))
+}
+
+func hash(event uint) uint {
+	h := event
+	h ^= h >> 16
+	h *= 0x119de1f3
+	h ^= h >> 15
+	h *= 0x119de1f3
+	h ^= h >> 16
+	return h
+}
```

Comment:
You provided a thorough explanation of this (and the rest of this file's code as well) during today's contributor sync. It would be very useful to have that explanation included as comments.

Reply:
Done in next push.  It's a big explanation, but it provides future devs with an appreciation of what's happening, and why it is needed (and if they read it carefully, what would be required in a modification).

Reply:
I may have to watch that recording to get the why :)

I'm finding it a bit hard to understand here how the logic specifically solves the problems we're looking at.

at the top it may help to type in some examples. you have some example already in the PR description- it would be helpful having this in comments in the code

Reply:
@amoore877 The "why" is to get a number that starts off repeatedly incrementing, and arrive at  a pseudo random hash that distributes through the entire number space.  That way, when we mod the "hash" by any other number, the remainders should be approximately evenly distributed through between [0, 1, 2, ... number).

To do so, you can't just multiply the number, as it will preserve its factors (divisible by 2 means always divisible by 2 because (2a) * b = 2(a * b).  So hashing algorithms tend to "mix" bits in from the number into the low end bits, hence all the >> 16 and >> 15.  

There's plenty of writeups on the internet about hashes.  https://en.wikipedia.org/wiki/Hash_function This is a non-cryptographic hash, which is the kind you would likely have to study when hand writing a hash table.  And this one isn't overly special, and like all non-cryptographic hashes, is designed to not use many CPU cycles to approximate a good dispersion.

So the multiplies and shifts combine to make something like 1, 2, 3, 4, 5 transform into 324231623, 01231385, 9923749123, 234202056, 49702374, which when modded by something should roughly distribute across the entire range of remainders, like (mod 3) (0, 2, 1, 1, 0, 2).  (All numbers above are examples, and not the actual output of this hash).

The reason we need it is because every bucket can potentially have a different size.  This means we want those remainders to distribute evenly across a large number of potential divisors.  If they didn't then "poll once over 4 polling periods" might look like (90%, 2% 5%, 3%) which would be a bad balance across that bucket.

Reply:
The code for this section has been removed.  I'm closing this conversation.